application-tour-ui/src/main/resources/TourCode/AjaxActions.xml

```diff
@@ -0,0 +1,536 @@
+  </object>
+  <content>{{velocity}}
+#**
+* Ajax Actions
+*
+* Author: Mohamed BOUSSAA &lt;mohamed.boussa@xwiki.com&gt; 
+*#
+
+#set($action = $request.getParameter("action"))
+#if ("$!action" != '')
+#if ($action == "editStep" &amp;&amp; "$!request.getParameter('BSTDocRef')" != '')
+  #set($docRef = $request.getParameter("BSTDocRef"))
+  #if ("$!docRef" != '')
+     #set($bootstrapTourDoc = $xwiki.getDocument($docRef))
+     #if ($bootstrapTourDoc)
+        ## Save step
+        #set($title = $request.getParameter("title"))
+        #set($path = $request.getParameter("path"))
+        #set($element = $request.getParameter("element"))
+        #set($backdrop = $request.getParameter("backdrop"))
+        #set($content = $request.getParameter("content"))
+        #set($placement = $request.getParameter("placement"))
+        #set($targetPage = $request.getParameter("targetPage"))
+        #set($nbSteps = $request.getParameter("nbSteps"))
+        #set($stepOrder = $request.getParameter("stepOrder"))
+        #set($isNewStep = true)
+        #if ("$!stepOrder" != '') ## Edit
+            #set($stepObj = $bootstrapTourDoc.getObject("TourCode.StepClass", "order", "$stepOrder"))
+            #if($stepObj)
+               #set($isNewStep = false)                
+            #end
+        #end
+        #if ($isNewStep)
+           ## New step
+           #set($stepObj = $bootstrapTourDoc.newObject('TourCode.StepClass'))
+        #end
+        #set($discard = $stepObj.set('path', $path))
+        #set($discard = $stepObj.set('element', $element))
+        #set($discard = $stepObj.set('title', $title))
+        #set($discard = $stepObj.set('placement', $placement))
+        #set($discard = $stepObj.set('targetPage', $targetPage))
+        #set($discard = $stepObj.set('backdrop', $backdrop))
+        #set($discard = $stepObj.set('content', $content))
+        #if ($isNewStep)
+           #set($discard = $stepObj.set('order', $nbSteps))
+        #end
+        #set($discard = $bootstrapTourDoc.save())
+        #set($BSTDocRef = $bootstrapTourDoc)
+        #set($pageMode = "edit")
+        {{include reference="TourCode.StepsListing"}}
+     #end
+  #end
+#elseif ($action == "deleteStep" &amp;&amp; "$!request.getParameter('BSTDocRef')" != '' &amp;&amp; "$!request.getParameter('stepOrder')" != '')
+   #set($docRef = $request.getParameter('BSTDocRef'))
+   #set($stepOrder =  $mathtool.toInteger($request.getParameter('stepOrder')))
+   #if ($xwiki.exists($docRef))
+      #set($bootstrapTourDoc = $xwiki.getDocument($docRef))
+      #set($stepObj = $bootstrapTourDoc.getObject("TourCode.StepClass", "order", "$stepOrder"))
+      #if ($stepObj)
+         #set($discard = $bootstrapTourDoc.removeObject($stepObj))
+         #set($discard = $bootstrapTourDoc.save())
+         ## Update other steps order
+         #set($hql = "select distinct step.order from Document doc, doc.object(TourCode.StepClass) as step where doc.fullName='$docRef' order by step.order")
```

Comment:
All the steps concerning the same tour are now stored in the same document, so this query is not needed anymore.